File: python/mpesa/callbacks.py
"""Callbacks Safaricom POSTs to CallBackURL -- STK family
(mirrors go/callbacks.go).

SECURITY POSTURE: these payloads carry NO HMAC signature -- anyone who
can reach your endpoint can forge one. Ranked controls: settle ONLY via
STK Query (ResultCode == "0") bound to YOUR original request record --
a body that parses is never proof of payment; cap request bodies in the
web framework (chars; >=1 MiB analogue of Go's ``http.MaxBytesReader``)
before :meth:`from_json`; gate the endpoint with callback_token.py's
URL tokens; binding on ``checkout_request_id`` is a DEDUP/IDEMPOTENCY
control, NOT origin authentication (a forged body carries whatever IDs
its author likes); treat all metadata values as ADVISORY ONLY (a forged
first item is indistinguishable from genuine); classify per
ADR-010-m-pesa-adapter.md (INDETERMINATE outcomes may still settle
minutes later).

Go-divergence footnote: stdlib ``json.loads`` ACCEPTS NaN/Infinity
literals Go's encoder rejects; typed helpers refuse them by gate.

Usage::

    result = StkCallbackResult.from_json(raw_body)
    order = repo.by_checkout(result.checkout_request_id)   # dedup FIRST
    if order is None:
        return                                             # unknown: ACK, ignore
    if result.classify() is ResultClass.SUCCESS:           # HINT only --
        resp = client.stk_query(STKQueryRequest(           # settle via the query
            checkout_request_id=result.checkout_request_id))
        if resp.result_code == "0":
            settle(order, receipt=result.mpesa_receipt(), amount=result.amount())
    else:
        mark_pending_reconcile(order)   # stk_query decides; never refund on a hit
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from .classification import ResultClass, classify_result_code
from .coercion import coerce_amount, coerce_int, coerce_str, first_wins, safe_json_int

__all__ = ["StkCallbackResult", "MetadataItem"]

# Ingestion cap in BYTES (Go maxResponseLen / TS MAX_RESPONSE_LEN parity).
# str bodies are measured as UTF-8 bytes: 1M CJK chars ~= 3 MiB and must
# not bypass the bound. See _check_body_size().
_MAX_BODY_BYTES = 1_048_576


def _check_body_size(data: "bytes | bytearray | str") -> None:
    """Reject bodies over the ingestion cap, measured in UTF-8 bytes.

    bytes input is measured directly; str input is measured as
    ``len(data.encode("utf-8"))`` so multi-byte payloads cannot smuggle
    up to 4x the intended bound past a char-count check.
    """
    size = len(data) if isinstance(data, (bytes, bytearray)) else \
        len(data.encode("utf-8"))
    if size > _MAX_BODY_BYTES:
        raise ValueError(
            f"mpesa: callback body exceeds {_MAX_BODY_BYTES} bytes")


def _item_name(entry: dict[str, Any]) -> str:
    """Name extraction surviving hostile values (unnameable -> "")."""
    value = entry.get("Name", "")
    if value is None or isinstance(value, str):
        return value if isinstance(value, str) else ""
    try:
        return str(value)
    except (ValueError, TypeError):
        return ""


@dataclass(frozen=True)
class MetadataItem:
    """One ``CallbackMetadata.Item`` entry; Value kept as the decoded
    JSON value (int/float/str/bool/None)."""

    name: str
    value_raw: Any = None


@dataclass(frozen=True)
class StkCallbackResult:
    """The ``Body.stkCallback`` outcome; metadata accessors tolerate its
    absence on failures. ``result_code`` normalized to str."""

    merchant_request_id: str = ""
    checkout_request_id: str = ""
    result_code: str = ""
    result_desc: str = ""
    _items: tuple[MetadataItem, ...] = field(default=(), repr=False)

    @classmethod
    def from_json(cls, data: "dict | bytes | str") -> "StkCallbackResult":
        """Parse the full envelope. Loud about SHAPE (missing keys raise
        ValueError naming them), tolerant about TYPES and about absent
        CallbackMetadata. Bodies over 1 MiB (UTF-8 bytes) are rejected
        before parsing."""
        if isinstance(data, (bytes, bytearray, str)):
            _check_body_size(data)
        if isinstance(data, (bytes, bytearray)):
            data = data.decode("utf-8", errors="replace")
        if isinstance(data, str):
            try:
                data = json.loads(data, parse_int=safe_json_int)
            except (ValueError, RecursionError) as exc:
                raise ValueError(
                    f"mpesa: unparseable callback body "
                    f"({type(exc).__name__})") from None
        if not isinstance(data, dict):
            raise ValueError("mpesa: unexpected STK callback shape: "
                             "missing or malformed Body")
        body = data.get("Body")
        inner = body.get("stkCallback") if isinstance(body, dict) else None
        if not isinstance(inner, dict):
            raise ValueError("mpesa: unexpected STK callback shape: "
                             "missing or malformed Body.stkCallback")
        wire_keys = (("merchant_request_id", "MerchantRequestID"),
                     ("checkout_request_id", "CheckoutRequestID"),
                     ("result_code", "ResultCode"),
                     ("result_desc", "ResultDesc"))
        missing = [key for _, key in wire_keys if key not in inner]
        if missing:
            raise ValueError(f"mpesa: unexpected STK callback shape: "
                             f"missing {', '.join(missing)}")
        meta = inner.get("CallbackMetadata") or {}
        item_list = meta.get("Item") if isinstance(meta, dict) else None
        items = tuple(MetadataItem(name=_item_name(e),
                                   value_raw=e.get("Value"))
                      for e in (item_list or []) if isinstance(e, dict))
        return cls(_items=items,
                   **{attr: coerce_str(inner[key]) or ""
                      for attr, key in wire_keys})

    def items(self) -> tuple[MetadataItem, ...]:
        """Raw metadata items, duplicates included, order preserved."""
        return self._items

    def duplicate_keys(self) -> int:
        """Count of items shadowed by an earlier same-named item."""
        seen: set[str] = set()
        return sum(1 for item in self._items
                   if item.name in seen or seen.add(item.name))

    def metadata(self) -> dict[str, Any]:
        """Flatten items FIRST-WINS; absent metadata yields {}. Example::

            md = result.metadata()   # {'Amount': 1.0, ...}
        """
        return first_wins([(i.name, i.value_raw) for i in self._items])

    def classify(self) -> ResultClass:
        """ADR-010 bucket for ``result_code`` (never auto-fail unknowns)."""
        return classify_result_code(self.result_code)

    def amount(self) -> float | None:
        """Amount as float via shared :func:`coerce_amount`; bools,
        >2**53 ints (precision), non-finite floats and non-decimal
        strings all yield None."""
        return coerce_amount(self._lookup("Amount"))

    def mpesa_receipt(self) -> str | None:
        """M-PESA receipt string, or None when absent."""
        value = self._lookup("MpesaReceiptNumber")
        return coerce_str(value)

    def transaction_date(self) -> int | None:
        """YYYYMMDDHHMMSS stamp as int; string-encoded accepted via
        strict coercion; bools rejected."""
        value = self._lookup("TransactionDate")
        if isinstance(value, bool):
            return None
        if isinstance(value, int):
            return value
        return coerce_int(value) if isinstance(value, str) else None

    def phone_number(self) -> str | None:
        """Payer MSISDN as ASCII-digit string; numeric encodings are
        stringified, hostile magnitudes decode to None via the parse_int
        hook, non-digit/non-ASCII refused."""
        value = self._lookup("PhoneNumber")
        if isinstance(value, bool):
            return None
        text = value.strip() if isinstance(value, str) else (
            str(value) if isinstance(value, int) else "")
        return text if text.isascii() and text.isdigit() else None

    def _lookup(self, name: str) -> Any:
        """First raw item value under *name*, else None."""
        for item in self._items:
            if item.name == name:
                return item.value_raw
        return None

File: go/callbacks.go
// Callbacks: payloads Safaricom POSTs to CallBackURL (STK family).

package mpesa

import (
	"encoding/json"
	"fmt"
	"math"
	"strings"
)

// maxCallbackBodyBytes caps the accepted callback body size at 1 MiB to
// prevent memory exhaustion from oversized or malicious POSTs.
const maxCallbackBodyBytes = 1 << 20

// STKCallback is the top-level envelope POSTed to CallBackURL.
//
// Safaricom callbacks carry NO HMAC signature, so ingestion endpoints must
// harden themselves. Ranked controls: settle ONLY via STKQuery with
// ResultCode==0 bound to YOUR CheckoutRequestID record (a body that parses
// is never proof of payment — forged callbacks parse fine); gate the
// endpoint with callbacktoken.go's bearer-capability URL tokens;
// CheckoutRequestID binding is a DEDUP/IDEMPOTENCY control, not origin
// authentication — a forged body carries whatever IDs its author likes.
// Validate amount/phone against the original request and wrap bodies in
// http.MaxBytesReader (recommend >=1 MiB cap) before unmarshalling.
type STKCallback struct {
	Body STKCallbackBody `json:"Body"`
}

// STKCallbackBody wraps the inner stkCallback object.
type STKCallbackBody struct {
	STKCallback STKCallbackResult `json:"stkCallback"`
}

// STKCallbackResult is the transaction outcome. CallbackMetadata is absent on
// failures — parse defensively via MetadataMap.
type STKCallbackResult struct {
	MerchantRequestID string            `json:"MerchantRequestID"`
	CheckoutRequestID string            `json:"CheckoutRequestID"`
	ResultCode        FlexString        `json:"ResultCode"`
	ResultDesc        string            `json:"ResultDesc"`
	CallbackMetadata  *CallbackMetadata `json:"CallbackMetadata,omitempty"`
}

func (r STKCallbackResult) metadataItems() []MetadataItem {
	if r.CallbackMetadata == nil {
		return nil
	}
	return r.CallbackMetadata.Item
}

// MetadataMap flattens callback metadata items, tolerating absent metadata;
// integral values decode as int64, decimals as float64. On duplicate names
// the FIRST item wins — check DuplicateKeys to detect lossy collisions.
func (r STKCallbackResult) MetadataMap() map[string]any {
	out := make(map[string]any)
	for _, item := range r.metadataItems() {
		if _, exists := out[item.Name]; !exists {
			out[item.Name] = item.decodedValue()
		}
	}
	return out
}

// DuplicateKeys reports how many metadata items were shadowed by an earlier
// same-named item (Safaricom duplicates have been observed on retries).
func (r STKCallbackResult) DuplicateKeys() int {
	seen := make(map[string]bool)
	dupes := 0
	for _, item := range r.metadataItems() {
		if seen[item.Name] {
			dupes++
			continue
		}
		seen[item.Name] = true
	}
	return dupes
}

// ParseSTKCallback decodes a raw callback POST body into its outcome,
// accepting EITHER Safaricom's full envelope {"Body":{"stkCallback":…}}
// OR a bare stkCallback result object (fixtures, replayed captures).
// Shape problems are loud errors; absent CallbackMetadata stays tolerated —
// read values defensively via MetadataMap. Bodies carry NO signature: cap
// them upstream (http.MaxBytesReader ≥ 1 MiB). Binding on CheckoutRequestID
// against your own request record dedups/idempotency-guards processing —
// it is NOT origin authentication; settle via STKQuery before any
// irreversible action.
func ParseSTKCallback(body []byte) (*STKCallbackResult, error) {
	if len(body) > maxCallbackBodyBytes {
		return nil, fmt.Errorf("mpesa: callback body too large (%d bytes, max %d)", len(body), maxCallbackBodyBytes)
	}
	var envelope struct {
		Body *struct {
			StkCallback *STKCallbackResult `json:"stkCallback"`
		} `json:"Body"`
	}
	if err := json.Unmarshal(body, &envelope); err != nil {
		return nil, fmt.Errorf("mpesa: parse stk callback body: %w", err)
	}
	if envelope.Body != nil && envelope.Body.StkCallback != nil {
		return envelope.Body.StkCallback, nil
	}
	var bare STKCallbackResult
	if err := json.Unmarshal(body, &bare); err != nil {
		return nil, fmt.Errorf("mpesa: unexpected stk callback shape: "+
			"want {\"Body\":{\"stkCallback\":{…}}} or a bare result object: %w", err)
	}
	if bare.ResultCode.String() == "" && bare.MerchantRequestID == "" &&
		bare.CheckoutRequestID == "" {
		return nil, fmt.Errorf("mpesa: unexpected stk callback shape: " +
			"missing Body.stkCallback")
	}
	return &bare, nil
}

// CallbackMetadata holds the Item list Safaricom sends on success.
type CallbackMetadata struct {
	Item []MetadataItem `json:"Item,omitempty"`
}

// MetadataItem is one named metadata value; Value stays raw because Safaricom
// mixes numeric and string encodings.
type MetadataItem struct {
	Name  string          `json:"Name"`
	Value json.RawMessage `json:"Value,omitempty"`
}

// decodedValue coerces raw JSON leniently: empty/null → nil (explicit absence,
// never a fake zero); integers without ./e/E → int64 (preserves
// TransactionDate/PhoneNumber magnitudes); then float64, string, bool, and
// finally raw bytes for anything unrecognized.
func (m MetadataItem) decodedValue() any {
	raw := strings.TrimSpace(string(m.Value))
	if raw == "" || raw == "null" {
		return nil
	}
	var f float64
	if err := json.Unmarshal(m.Value, &f); err == nil {
		if !strings.ContainsAny(raw, ".eE") && f >= math.MinInt64 && f <= math.MaxInt64 {
			return int64(f)
		}
		return f
	}
	var s string
	if err := json.Unmarshal(m.Value, &s); err == nil {
		return s
	}
	var b bool
	if err := json.Unmarshal(m.Value, &b); err == nil {
		return b
	}
	return m.Value
}
